### PR TITLE
Kafka: provide new option to enable scaling past partition count

### LINKED ISSUE #1683: [KafkaScaler] Scale beyond the number of partitions
### Proposal

Add an option to disable the check that prevent KafkaScaler to scale beyond the number of partitions

### Use-Case

When KEDA is used to scale CRDs, the number of replicas does not necessarily match the number of Kafka consumers. For instance, KEDA can we used to increase/decrease the throughput (msgs/sec). In this usecase it makes sense to have more replicas than partitions.

### PR DESCRIPTION
**Pull request by lionelvillard**
_Provide a description of what has been changed_
 
### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated

Relates to #1683

### CODE REVIEW

**Comment by lionelvillard** [comment]
Here an initial proposal. If accepted, I'll add some tests and documentation.

File: pkg/scalers/kafka_scaler.go [hunk 1]
```diff
@@ -25,11 +25,12 @@ type kafkaScaler struct {
...
-	group             string
-	topic             string
-	lagThreshold      int64
-	offsetResetPolicy offsetResetPolicy
+	bootstrapServers       []string
+	group                  string
+	topic                  string
+	lagThreshold           int64
+	offsetResetPolicy      offsetResetPolicy
+	partitionLimitDisabled bool
```
Comment (zroubalik) [inline pkg/scalers/kafka_scaler.go:33]: Generally ok, but I don't like the name of the property, isn't there a better option?

**Comment by tomkerkhove** [comment]
Would you mind updating the docs please?

File: pkg/scalers/kafka_scaler.go (hunk 1 above)
Comment (zroubalik) [reply pkg/scalers/kafka_scaler.go:33]: @tomkerkhove do you have any opinion on the naming?

Comment (lionelvillard) [reply pkg/scalers/kafka_scaler.go:33]: I'm not a big fan of the naming either. Alternatives: `partitionCheckDisabled` or `allowIdlePartitions`

What about `allowIdlePartitions`? I got this from `Kafka, the definitive guide` which talks about idle partitions. It is also a common pattern to keep idle consumers for HA.

Comment (zroubalik) [reply pkg/scalers/kafka_scaler.go:33]: Sounds better imho, it should come with a solid docs though.

Comment (tomkerkhove) [reply pkg/scalers/kafka_scaler.go:33]: LGTM

Comment (lionelvillard) [reply pkg/scalers/kafka_scaler.go:33]: Ok great! I will update this PR and create a new one for the doc.
Thanks

Comment (lionelvillard) [reply pkg/scalers/kafka_scaler.go:33]: Oops sorry I meant to type `allowIdleConsumers`, which is more accurate. LMK if you object to the change.

**Comment by lionelvillard** [comment]
Doc PR: https://github.com/kedacore/keda-docs/pull/407

File: pkg/scalers/kafka_scaler.go [hunk 2]
```diff
@@ -360,9 +370,11 @@ func (s *kafkaScaler) GetMetrics(ctx context.Context, metricName string, metricS
 
 	kafkaLog.V(1).Info(fmt.Sprintf("Kafka scaler: Providing metrics based on totalLag %v, partitions %v, threshold %v", totalLag, len(partitions), s.metadata.lagThreshold))
 
-	// don't scale out beyond the number of partitions
-	if (totalLag / s.metadata.lagThreshold) > int64(len(partitions)) {
-		totalLag = int64(len(partitions)) * s.metadata.lagThreshold
+	if !s.metadata.allowIdleConsumers {
```
Comment (zroubalik) [inline pkg/scalers/kafka_scaler.go:373]:
```suggestion
	if s.metadata.allowIdleConsumers {
```

isn't this better? 😄

**Review by zroubalik — COMMENTED** [review]
could you please add some tests to check wheteher the value is parsed correctly?

**Comment by lionelvillard** [comment]
> Looking good, could you please add a test to check whether the value is parsed correctly?

Isn't this good enough: https://github.com/kedacore/keda/pull/1684/files#diff-74937abdfe4bcd7bfa4682524e865761aa8aa4b824a4fc2f6fe03bdea3a94e14R34 ?

**Comment by zroubalik** [comment]
> > Looking good, could you please add a test to check whether the value is parsed correctly?
> 
> Isn't this good enough: https://github.com/kedacore/keda/pull/1684/files#diff-74937abdfe4bcd7bfa4682524e865761aa8aa4b824a4fc2f6fe03bdea3a94e14R34 ?

Would be nice to add entry in here: https://github.com/kedacore/keda/pull/1684/files#diff-74937abdfe4bcd7bfa4682524e865761aa8aa4b824a4fc2f6fe03bdea3a94e14L46
Nothing complex is needed.

**Review by zroubalik — APPROVED** [review]
LGTM

**Pull request merged**